data_model/src/transaction.rs

```diff
@@ -287,22 +287,28 @@ impl From<VersionedValidTransaction> for VersionedTransaction {
     }
 }
 
-/// This structure represents transaction in non-trusted form.
-///
-/// `Iroha` and its' clients use [`Transaction`] to send transactions
-/// via network.  Direct usage in business logic is strongly
-/// prohibited. Before any interactions `accept`.
-#[version(n = 1, versioned = "VersionedTransaction")]
+/// Trait for signing transactions
+pub trait Sign {
+    /// Sign transaction with provided key pair.
+    ///
+    /// # Errors
+    ///
+    /// Fails if signature creation fails
+    fn sign(
+        self,
+        key_pair: iroha_crypto::KeyPair,
```

Comment:
Just an architectural question. Do we **need** to provide the public key inside the key pair. Knowing the algorithm we can derive it.

Reply:
Yes, we could implement `From<PrivateKey> for PublicKey` and recover `PublicKey` on demand.
In such case we don't need `KeyPair` at all.

```rust
impl From<PrivateKey> for PublicKey {
    #[allow(clippy::expect_used)]
    fn from(private_key: PrivateKey) -> Self {
        let digest_function = private_key.digest_function();
        let key_gen_option = Some(UrsaKeyGenOption::FromSecretKey(UrsaPrivateKey(
            private_key.payload,
        )));
        let (mut public_key, _) = match digest_function {
            Algorithm::Ed25519 => Ed25519Sha512.keypair(key_gen_option),
            Algorithm::Secp256k1 => EcdsaSecp256k1Sha256::new().keypair(key_gen_option),
            Algorithm::BlsNormal => BlsNormal::new().keypair(key_gen_option),
            Algorithm::BlsSmall => BlsSmall::new().keypair(key_gen_option),
        }
        .expect("can't fail for valid `PrivateKey`");
        PublicKey {
            digest_function: private_key.digest_function,
            payload: core::mem::take(&mut public_key.0),
        }
    }
}
```

Reply:
User can provide invalid private key and I don't think we need to panic in that case. Better to implement `TryFrom`, I believe

Reply:
`TryFrom` is good option, i just think that it may be redundant to return error in case if invariants for private keys are checked before.

Reply:
Hm, if this invariant is guaranteed, then I agree